cli/npm/resolution/graph.rs

```diff
@@ -1,198 +1,348 @@
 // Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
 
-use std::borrow::Cow;
-use std::collections::BTreeMap;
-use std::collections::BTreeSet;
+use std::collections::hash_map::DefaultHasher;
 use std::collections::HashMap;
+use std::collections::HashSet;
 use std::collections::VecDeque;
+use std::hash::Hash;
+use std::hash::Hasher;
 use std::sync::Arc;
 
 use deno_core::anyhow::bail;
 use deno_core::anyhow::Context;
 use deno_core::error::AnyError;
-use deno_core::futures;
 use deno_core::parking_lot::Mutex;
-use deno_core::parking_lot::MutexGuard;
+use deno_graph::npm::NpmPackageNv;
 use deno_graph::npm::NpmPackageReq;
-use deno_graph::semver::Version;
 use deno_graph::semver::VersionReq;
 use log::debug;
-use once_cell::sync::Lazy;
 
-use crate::npm::cache::should_sync_download;
 use crate::npm::registry::NpmDependencyEntry;
 use crate::npm::registry::NpmDependencyEntryKind;
 use crate::npm::registry::NpmPackageInfo;
 use crate::npm::registry::NpmPackageVersionInfo;
+use crate::npm::resolution::common::resolve_best_package_version_and_info;
+use crate::npm::resolution::snapshot::SnapshotPackageCopyIndexResolver;
 use crate::npm::NpmRegistryApi;
 
+use super::common::version_req_satisfies;
+use super::common::LATEST_VERSION_REQ;
 use super::snapshot::NpmResolutionSnapshot;
-use super::snapshot::SnapshotPackageCopyIndexResolver;
-use super::NpmPackageNodeId;
+use super::NpmPackageId;
 use super::NpmResolutionPackage;
 
-pub static LATEST_VERSION_REQ: Lazy<VersionReq> =
-  Lazy::new(|| VersionReq::parse_from_specifier("latest").unwrap());
-
-/// A memory efficient path of visited name and versions in the graph
-/// which is used to detect cycles.
-///
-/// note(dsherret): although this is definitely more memory efficient
-/// than a HashSet, I haven't done any tests about whether this is
-/// faster in practice.
-#[derive(Default, Clone)]
-struct VisitedVersionsPath {
-  previous_node: Option<Arc<VisitedVersionsPath>>,
-  visited_version_key: String,
+/// A unique identifier to a node in the graph.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Ord, PartialOrd)]
+struct NodeId(u32);
+
+/// Reference to the node above this one in the graph.
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Ord, PartialOrd)]
+enum NodeParent {
+  /// This means the node is at the top of the graph and is specified
+  /// in Deno code.
+  Root(NpmPackageNv),
```

Comment:
Question: isn't it very costly to potentially clone this info multiple times? Should we consider in the future to use some kind of `Rc`/`Arc` to avoid expensive clones?

Reply:
I'm able to remove `NodeParent` completely actually. Will update in a second.